File: src/main/java/julius/game/chessengine/evaluation/nn/NNUEFeatures.java
package julius.game.chessengine.evaluation.nn;

/**
 * HalfKP feature index computation for NNUE evaluation.
 *
 * <p>Feature scheme: for each perspective (white/black), one binary feature is
 * active per non-king piece on the board. The feature index encodes:
 * {@code kingSquare * NUM_PIECE_FEATURES + pieceIndex * 64 + pieceSquare}
 *
 * <p>Piece index mapping (10 piece types: 5 types x 2 colors):
 * <pre>
 *   0 = own pawn        5 = enemy pawn
 *   1 = own knight      6 = enemy knight
 *   2 = own bishop      7 = enemy bishop
 *   3 = own rook        8 = enemy rook
 *   4 = own queen       9 = enemy queen
 * </pre>
 *
 * <p>"Own" and "enemy" are relative to the perspective side. Feature indices
 * are symmetric: what white sees as "own pawn on e4" is different from what
 * black sees. For black's perspective, squares are mirrored vertically
 * (square XOR 56) so the network sees a consistent orientation.
 */
public final class NNUEFeatures {

    private NNUEFeatures() {}

    /** Number of piece types per side (pawn, knight, bishop, rook, queen). Kings excluded. */
    public static final int PIECE_TYPES = 5;
    /** Number of piece-color combinations (own 5 + enemy 5). */
    public static final int PIECE_INDICES = 10;
    /** Features per king square: 10 piece indices * 64 squares. */
    public static final int FEATURES_PER_KING = PIECE_INDICES * 64;  // 640
    /** Total feature space per perspective: 64 king squares * 640. */
    public static final int TOTAL_FEATURES = 64 * FEATURES_PER_KING; // 40960
    /** Size of the first hidden layer (accumulator width). */
    public static final int ACCUMULATOR_SIZE = 128;

    /**
     * Compute the HalfKP feature index for a piece from a given perspective.
     *
     * @param kingSq     king square of the perspective side (0-63, a1=0, h8=63)
     * @param pieceSq    square of the piece (0-63)
     * @param pieceType  piece type bits (1=pawn..5=queen). Kings are excluded.
     * @param pieceWhite true if the piece is white
     * @param perspWhite true if computing from white's perspective
     * @return feature index in [0, TOTAL_FEATURES), or -1 if invalid
     */
    public static int featureIndex(int kingSq, int pieceSq, int pieceType, boolean pieceWhite, boolean perspWhite) {
        if (pieceType < 1 || pieceType > 5) return -1; // skip kings and invalid

        // For black perspective, mirror all squares vertically
        int ks = perspWhite ? kingSq : (kingSq ^ 56);
        int ps = perspWhite ? pieceSq : (pieceSq ^ 56);

        // Piece index: 0-4 = own pieces, 5-9 = enemy pieces
        boolean isOwn = (pieceWhite == perspWhite);
        int pieceIdx = (pieceType - 1) + (isOwn ? 0 : PIECE_TYPES);

        return ks * FEATURES_PER_KING + pieceIdx * 64 + ps;
    }

    /**
     * Maximum number of active features per perspective (at most 15 non-king pieces
     * can be on the board: 8 pawns + 2N + 2B + 2R + 1Q, and opponent's pieces too).
     * In practice ~20-30 features active.
     */
    public static final int MAX_ACTIVE_FEATURES = 30;
}

File: scripts/nn-training/nnue_features.py
#!/usr/bin/env python3
"""
HalfKP feature index computation for NNUE training.
Must produce identical indices to Java NNUEFeatures.java.
"""
import chess

PIECE_TYPES = 5       # pawn, knight, bishop, rook, queen (no king)
PIECE_INDICES = 10    # own 5 + enemy 5
FEATURES_PER_KING = PIECE_INDICES * 64  # 640
TOTAL_FEATURES = 64 * FEATURES_PER_KING  # 40960
ACCUMULATOR_SIZE = 256

# Piece type mapping: chess.PAWN=1..chess.QUEEN=5 (matching Java MoveHelper)
PIECE_TYPE_MAP = {
    chess.PAWN: 1,
    chess.KNIGHT: 2,
    chess.BISHOP: 3,
    chess.ROOK: 4,
    chess.QUEEN: 5,
}


def feature_index(king_sq, piece_sq, piece_type_bits, piece_is_white, persp_is_white):
    """Compute HalfKP feature index. Must match Java NNUEFeatures.featureIndex()."""
    if piece_type_bits < 1 or piece_type_bits > 5:
        return -1

    # For black perspective, mirror squares vertically
    ks = king_sq if persp_is_white else (king_sq ^ 56)
    ps = piece_sq if persp_is_white else (piece_sq ^ 56)

    # Own vs enemy
    is_own = (piece_is_white == persp_is_white)
    piece_idx = (piece_type_bits - 1) + (0 if is_own else PIECE_TYPES)

    return ks * FEATURES_PER_KING + piece_idx * 64 + ps


def extract_active_features(board):
    """
    Extract active HalfKP feature indices for both perspectives.
    Returns (white_features, black_features) as lists of ints.
    """
    wk = board.king(chess.WHITE)
    bk = board.king(chess.BLACK)

    white_features = []
    black_features = []

    for sq in range(64):
        piece = board.piece_at(sq)
        if piece is None or piece.piece_type == chess.KING:
            continue

        pt_bits = PIECE_TYPE_MAP.get(piece.piece_type)
        if pt_bits is None:
            continue

        is_white = piece.color == chess.WHITE

        # White perspective
        idx_w = feature_index(wk, sq, pt_bits, is_white, True)
        if idx_w >= 0:
            white_features.append(idx_w)

        # Black perspective
        idx_b = feature_index(bk, sq, pt_bits, is_white, False)
        if idx_b >= 0:
            black_features.append(idx_b)

    return white_features, black_features


def test_symmetry():
    """Quick sanity check: symmetric positions should produce mirrored features."""
    board = chess.Board()
    wf, bf = extract_active_features(board)
    print(f"Starting position: {len(wf)} white features, {len(bf)} black features")
    assert len(wf) == 30, f"Expected 30, got {len(wf)}"  # 16 pawns + 4N + 4B + 4R + 2Q
    assert len(bf) == 30, f"Expected 30, got {len(bf)}"
    print("Symmetry test passed!")


if __name__ == "__main__":
    test_symmetry()

    # Test a specific position
    board = chess.Board("rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1")
    wf, bf = extract_active_features(board)
    print(f"After 1.e4: {len(wf)} white features, {len(bf)} black features")
    print(f"  White features (first 5): {wf[:5]}")
    print(f"  Black features (first 5): {bf[:5]}")
